legate/legate_kvikio/core.py

```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2023, NVIDIA CORPORATION. All rights reserved.
+# See file LICENSE for terms.
+
+import struct
+from enum import IntEnum
+from typing import Any
+
+import numpy as np
+
+import legate.core.types as types
+from legate.core import Store, get_legate_runtime
+
+from .library import user_context, user_lib
+
+
+class TaskOpCode(IntEnum):
+    WRITE = user_lib.cffi.OP_WRITE
+    READ = user_lib.cffi.OP_READ
+
+
+def _get_legate_store(input: Any) -> Store:
+    """Extracts a Legate store from any object
+       implementing the legate data interface
+
+    Args:
+        input (Any): The input object
+
+    Returns:
+        Store: The extracted Legate store
+    """
+    if isinstance(input, Store):
+        return input
+    data = input.__legate_data_interface__["data"]
+    field = next(iter(data))
+    array = data[field]
+    _, store = array.stores()
+    return store
+
+
+def zero() -> Store:
+    """Create a Legates store representing a single zero scalar
+
+    Returns:
+        Store: A Legate store representing a scalar zero
+    """
+    data = bytearray(8)
+    buf = struct.pack(f"{len(data)}s", data)
+    future = get_legate_runtime().create_future(buf, len(buf))
+    return user_context.create_store(
+        types.int64,
+        shape=(1,),
+        storage=future,
+        optimize_scalar=True,
+    )
+
+
+def get_written_nbytes(input: Store) -> int:
+    """Blocking call to get the value of `write()`
+
+    Args:
+        input (Store): The Legate store encapsulating a scalar
+
+    Returns:
+        float: A Python scalar
+    """
+    buf = input.storage.get_buffer(np.int64().itemsize)
+    result = np.frombuffer(buf, dtype=np.int64, count=1)
+    return int(result[0])
+
+
+def write(path: str, obj: Any) -> Store:
+    """Write data to disk using KvikIO
+
+    Args:
+        obj (Any): A Legate store or any object implementing
+                     the Legate data interface.
+    """
+    input = _get_legate_store(obj)
+    task = user_context.create_auto_task(TaskOpCode.WRITE)
+    task.add_scalar_arg(path, types.string)
+    task.add_input(input)
+    task.set_side_effect(True)
+    # TODO: return a handle instead.
+    output = zero()
+    task.add_reduction(output, types.ReductionOp.ADD)
+    task.execute()
+    return output
+
+
+def read(path: str, obj: Any) -> None:
+    """Read data from disk using KvikIO
+
+    Args:
+        obj (Any): A Legate store or any object implementing
+                     the Legate data interface.
+    """
+    output = _get_legate_store(obj)
+    task = user_context.create_auto_task(TaskOpCode.READ)
```

Comment:
I think this task would be more appropriate as an unbounded-output task, similar to how `unique` works in cuNumeric https://github.com/nv-legate/cunumeric/blob/branch-23.05/cunumeric/deferred.py#L3426.

This would also change the interface, such that the output store is created inside the function call, and returned as the result (rather than having the user pre-allocate a store of appropriate size).

Alternatively, if possible you could inspect the file's metadata, and preallocate an appropriate store inside this function, before doing the actual read. You wouldn't need to use unbounded stores in that case.

I believe for safety you also want to make it clear to legate that this must be launched as a singleton task:

```
    task = user_context.create_auto_task(TaskOpCode.READ)
    task.add_scalar_arg(path, types.string)
    task.add_output(output)
    # the whole output Store must be accessible by all point tasks
    # Stores accessed with write permissions cannot be shared
    # therefore the launch can only contain one point task.
    task.add_broadcast(output)  
    task.execute()
```

Reply:
> I think this task would be more appropriate as an unbounded-output task, similar to how `unique` works in cuNumeric https://github.com/nv-legate/cunumeric/blob/branch-23.05/cunumeric/deferred.py#L3426.
> 
> This would also change the interface, such that the output store is created inside the function call, and returned as the result (rather than having the user pre-allocate a store of appropriate size).

In this case, I would still have to know the total size of the file beforehand in order for each task to determine their file offset, right? 

> Alternatively, if possible you could inspect the file's metadata, and preallocate an appropriate store inside this function, before doing the actual read. You wouldn't need to use unbounded stores in that case.

Agree, the user API could support a default that infer the buffer type. However, the plan is to incorporate this into the existing KvikIO API so that something like the following will just work, no matter if `buf` is a NumPy, CuPy, or cuNumeric array:
```python
f = kvikio.CuFile("test-file", "r")
f.read(buf)
```  

> I believe for safety you also want to make it clear to legate that this must be launched as a singleton task:

I don't think this should be a singleton task, I want Legate to run multiple tasks in parallel. I am hoping that by calculating the offset based on the input store, each task will read its non-overlapping part of the file?
```c++
    auto shape = store.shape<1>();
    auto acc = store.read_accessor<char, 1>();
    size_t strides[1];
    const char* data = acc.ptr(shape, strides);
    size_t itemsize  = sizeof_legate_type_code(store.code());
    assert(strides[0] == itemsize);  // Must be contiguous
    size_t nbytes = shape.volume() * itemsize;  
    size_t offset = shape.lo.x * itemsize;  // Offset in bytes
```

Reply:
> In this case, I would still have to know the total size of the file beforehand in order for each task to determine their file offset, right?

Not necessarily. Each point task could query the size of the file independently and start reading at `size * point_task_id / num_point_tasks`. Each task would need to return to the caller how many elements it actually read, so the returned buffers can be virtually "stitched together". See the code in `unique` for how that's done (the term "weights" refers to how many elements each point task read).

> Agree, the user API could support a default that infer the buffer type. However, the plan is to incorporate this into the existing KvikIO API

Then I assume there's calls so that the user code can query the size of the file, so it can pre-allocate an array of the right size?

> I want Legate to run multiple tasks in parallel. I am hoping that by calculating the offset based on the input store, each task will read its non-overlapping part of the file?

Yes, that should work.

Reply:
Thanks for the clarification @manopapad